### ISSUE TITLE
FileDocumentStore cannot be used as IDocumentStore

### ISSUE THREAD

**Issue by TFleury** [OrchardCMS/OrchardCore#12910]
### Scenario
If you want to use `FileDocumentStore` as default `IDocumentStore` implementation instead of YesSql DocumentStore.

### To Reproduce
```cs
.AddDataAccess()
        .ConfigureServices(services => services
            // Use FileDocumentStore as IDocumentStore
            .RemoveAll<IDocumentStore>()
            .AddScoped<IDocumentStore>(sp => sp.GetRequiredService<IFileDocumentStore>())
        )
```

### Expected behavior
We should be able to replace `DocumentStore` with `FileDocumentStore`.

### Current behavior
The `UpdateAsync` method of `FileDocumentStore` saves the document inside `IDocumentStore.AfterCommitSuccess` which leads to an infinite loop if you use `FileDocumentStore` as `IDocumentStore` implementation.

**Comment by sebastienros** [comment]
Looking back at the original issue, here is how Placement is stored in files instead of the database.

```c#
    [Feature("OrchardCore.Placements.FileStorage")]
    public class FileContentDefinitionStartup : StartupBase
    {
        public override void ConfigureServices(IServiceCollection services)
        {
            services.RemoveAll<IPlacementStore>();
            services.AddScoped<IPlacementStore, FilePlacementsStore>();
        }
    }
```

Is the goal here to hijack all the things that can store a document to use the filesystem instead?

**Comment by TFleury** [comment]
> Is the goal here to hijack all the things that can store a document to use the filesystem instead?

Yes, that's it. Not really for a CMS usage, but more to use core features as an application framework where you don't want a document database.

**Comment by sebastienros** [comment]
Not your fault but there are already too many interfaces IMO. 

The following is general, not specific to this PR: There should be a single interface, `IDocumentStore` and then concrete implementations. You want to explicitly store in the filesystem? then resolve the implementation, not the interface. Now with keyed services we can even decide to register different implementation of the same interface per consumer. For instance the Placement manager could try to resolve the `IDocumentStore` for the `Placement` class, and the content definition one could resolve it for `ContentTypeDefinition`. This way we don't need to create different interfaces for each case, and developers could just register the implementation they want to override like this: `services.AddScoped<IDocumentStore, FileSystemStore>(key: typeof(Placement));`

**Comment by TFleury** [comment]
I agree to use keyed services instead of duplicate interfaces, but if you want to use document type as key, you will need one registration per document type, and you will have one instance of DocumentStore per document type, what is not expected.
It would rather be `services.AddKeyedScoped<IDocumentStore, FileSystemDocumentStore>("file")`

`IDocumentStore` is mainly used by `DocumentManager`, so in that case we need to adapt the way `IDocumentStore` is resolved.

`IFileDocumentStore` exists since v1.0.0, it should be deprecated before being removed.

**Comment by TFleury** [comment]
I thought about it for a few hours and don't see a clean way to get rid of `IFileDocumentStore` without introducing a new interface to select `IDocumentStore` per document type. So it would remove an interface to add another and also introduce breaking changes.
